app/src/main/java/com/devindi/wallpaper/model/map/MapImageGenerator.kt

```diff
@@ -1,13 +1,91 @@
 package com.devindi.wallpaper.model.map
 
 import android.graphics.Bitmap
+import android.graphics.Canvas
+import com.devindi.wallpaper.misc.RectD
+import com.google.firebase.perf.metrics.AddTrace
+import org.osmdroid.util.MapTileIndex
+import timber.log.Timber
 
 /**
  * Generates bitmap by params (source, center coordinates, zoom, size)
  */
-class MapImageGenerator() {
+class MapImageGenerator(
+    private val tileProvider: SyncMapTileProvider,
+    private val sourceFactory: TileSourceFactory,
+    private val tileUtils: TileUtils
+) {
 
-    fun generate(): Bitmap {
-        throw NotImplementedError()
+    fun generate(mapSourceId: String, zoom: Int, centerLat: Double, centerLong: Double, width: Int, height: Int): Bitmap {
+        Timber.d("Generate bitmap: $centerLat, $centerLong, $width, $height")
+        val tileSource = sourceFactory.getTileSource(mapSourceId)
+        val tilesRect = tileUtils.getTilesRect(centerLat, centerLong, width, height, tileSource.tileSizePixels, zoom)
```

Comment:
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/devindi/mapwall/blob/63cb55c913b0d690492f74a58d6634a0c41df234/app/src/main/java/com/devindi/wallpaper/model/map/MapImageGenerator.kt#L22"></span>Exceeded max line length (100) (cannot be auto-corrected)</td>
    </tr>
  </tbody>
</table>